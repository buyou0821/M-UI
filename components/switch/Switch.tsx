import React, { forwardRef, useRef, useState, useLayoutEffect } from 'react';
import clsx from 'clsx';
import { usePrefixCls } from '../_util/hooks';
import { Omit } from '../_util/type';
import { useForkRef } from '../_util/reactHelpers';
import { tuple } from '../_util/type';
import Button from '../button';

const SwitchColors = tuple('default', 'primary', 'secondary', 'success', 'warning', 'danger');
export type SwitchColor = (typeof SwitchColors)[number];

export interface SwitchEvent {
  target: {
    type: 'switch';
    checked: boolean;
  } & SwitchProps;
  preventDefault(): void;
  stopPropagation(): void;
}
export type SwitchSize = 'default' | 'small';
export type SwitchChangeEventHandler = (checked: boolean, event: SwitchEvent) => void;
export type SwitchClickEventHandler = SwitchChangeEventHandler;

export interface SwitchProps
  extends Omit<React.DOMAttributes<HTMLButtonElement>, 'onChange' | 'onClick'> {
  disabled?: boolean;
  checked?: boolean;
  autoFocus?: boolean;
  loading?: boolean;
  size?: SwitchSize;
  onChange?: SwitchChangeEventHandler;
  onClick?: SwitchClickEventHandler;
  style?: React.CSSProperties;
  color?: SwitchColor;
}

const Switch = forwardRef((props: SwitchProps, ref: React.RefObject<HTMLButtonElement>) => {
  const {
    checked = false,
    autoFocus,
    disabled,
    onChange,
    onClick,
    color = 'default',
    ...rest
  } = props;
  const [isChecked, setIsChecked] = useState<boolean>(checked);
  const prefixCls = usePrefixCls('switch');
  const switchRef = useRef<HTMLButtonElement>(null);
  const handleRef = useForkRef(ref, switchRef);

  useLayoutEffect(() => {
    if (autoFocus && !disabled) {
      focus();
    }
  }, [autoFocus, disabled]);

  const focus = () => {
    if (switchRef && switchRef.current) {
      switchRef.current.focus();
    }
  };

  const blur = () => {
    if (switchRef.current) {
      switchRef.current.blur();
    }
  };

  const handleMouseUp = (event: React.MouseEvent<HTMLButtonElement>) => {
    const { onMouseUp } = props;
    blur();
    if (onMouseUp) {
      onMouseUp(event);
    }
  };

  const handleChange = (checked: boolean, evt: SwitchEvent) => {
    if (disabled) return;
    if (!('checked' in props)) {
      setIsChecked(checked);
    }
    if (onChange) {
      onChange(checked, evt);
    }
  };

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    const newChecked = !isChecked;
    const evt: SwitchEvent = {
      target: {
        type: 'switch',
        checked: newChecked,
        ...props,
      },
      preventDefault() {
        event.preventDefault();
      },
      stopPropagation() {
        event.stopPropagation();
      },
    };
    handleChange(newChecked, evt);
    if (onClick) {
      onClick(newChecked, evt);
    }
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLButtonElement>) => {
    const newChecked = {
      37: false, // Left
      39: true, // Right
    }[event.keyCode];
    const evt: SwitchEvent = {
      target: {
        type: 'switch',
        checked: newChecked,
        ...props,
      },
      preventDefault() {
        event.preventDefault();
      },
      stopPropagation() {
        event.stopPropagation();
      },
    };
    handleChange(newChecked, evt);
  };

  const classes = clsx(prefixCls, {
    [`${prefixCls}--checked`]: isChecked,
    [`${prefixCls}--${color}`]: color && !disabled && isChecked,
  });

  const rippleColor = color && !disabled && isChecked ? color : 'default';

  return (
    <button
      {...rest}
      ref={handleRef}
      className={classes}
      role="switch"
      aria-checked={isChecked}
      onMouseUp={handleMouseUp}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
    >
      <span className={`${prefixCls}__node`} />
      <Button
        component="span"
        shape="icon"
        center
        className={`${prefixCls}__ripple`}
        disabled={disabled}
        color={rippleColor}
      >
        <span className={`${prefixCls}__circle`} />
      </Button>
    </button>
  );
});

export default Switch;
