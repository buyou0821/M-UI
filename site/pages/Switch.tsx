import React from 'react';
import { Switch } from 'components';

export default () => {
  return (
    <div style={{ padding: 20 }}>
      <Switch
        autoFocus
        onClick={(checked, e) => {
          console.log(checked, e);
        }}
      />
      <Switch disabled color="primary" />
      <Switch color="secondary" />
    </div>
  );
};
