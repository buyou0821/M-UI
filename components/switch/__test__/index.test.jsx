import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Switch from '..';

describe('Switch', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<Switch />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should support checked', () => {
    render(<Switch checked />);
    expect(document.querySelectorAll('.ty-switch--checked').length).toBe(1);
  });

  it('should support keyBoard events', () => {
    render(<Switch autoFocus />);
    const switchDOM = document.querySelector('.ty-switch');
    expect(switchDOM.classList).not.toContain('ty-switch--checked');
    fireEvent.keyDown(switchDOM, { key: 'right', keyCode: 39 });
    expect(switchDOM.classList).toContain('ty-switch--checked');
    fireEvent.keyDown(switchDOM, { key: 'left', keyCode: 37 });
    expect(switchDOM.classList).not.toContain('ty-switch--checked');
  });

  it('should fire onChange correctly', () => {
    const handleChange = jest.fn();
    render(<Switch onChange={handleChange} />);
    const switchDOM = document.querySelector('.ty-switch');
    fireEvent.click(switchDOM);
    expect(handleChange).toHaveBeenCalled();
  });

  it('should support disabled', () => {
    const handleChange = jest.fn();
    const { container } = render(<Switch disabled onChange={handleChange} />);
    fireEvent.click(container);
    expect(handleChange).not.toHaveBeenCalled();
  });

  it('should fire onClick correctly', () => {
    const handleClick = jest.fn();
    render(<Switch onClick={handleClick} />);
    const switchDOM = document.querySelector('.ty-switch');
    fireEvent.click(switchDOM);
    expect(handleClick).toHaveBeenCalled();
  });

  it('should support defaultChecked', () => {
    const { container } = render(<Switch defaultChecked />);
    expect(container.querySelectorAll('.ty-switch--checked').length).toBe(1);
    fireEvent.click(container.querySelector('.ty-switch'));
    expect(container.querySelectorAll('.ty-switch--checked').length).toBe(0);
  });
});
