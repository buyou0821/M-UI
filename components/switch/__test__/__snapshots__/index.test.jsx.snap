// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render correctly 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="ty-switch"
    role="switch"
  >
    <span
      class="ty-switch__node"
    />
    <span
      class="ty-ripple ty-btn ty-switch__ripple ty-btn__icon ty-btn__icon--default"
    >
      <span
        class="ty-btn__wrapper"
      >
        <span
          class="ty-switch__circle"
        />
      </span>
      <span
        class="ty-ripple__wrapper"
      />
    </span>
  </button>
</DocumentFragment>
`;
